Complete a majority of the rest of DI

Nokel81 opened the pull request:
- All KubeApi and KubeObjectStore's
- All of Catalog
- ApiManager
- KubeObjectDetailRegistry
- All BaseStore's and their migrations
- LensProxy and Router
- WindowManager
- App Menu
- Tray Icon
- ThemeStore



Know Issues

- [ ] Catalog Menu doesn't change the view
- [ ] Cluster overview metrics don't seem to load
- [ ] Namespaced resources don't load

fixes https://github.com/lensapp/lens/issues/4632
fixes https://github.com/lensapp/lens/issues/4712
fixes https://github.com/lensapp/lens/issues/4713
fixes https://github.com/lensapp/lens/issues/4633
fixes https://github.com/lensapp/lens/issues/4714
fixes https://github.com/lensapp/lens/issues/4715
fixes https://github.com/lensapp/lens/issues/4716

fixes all the kube api and kube stores

jansav commented on src/common/getDiForUnitTesting.ts:
```diff
@@ -0,0 +1,39 @@
+/**
```
I don't think we need this. Everything under `common` have expectation that it works with either `mainDi` or `rendererDi` so we could use either of those in these tests. Or even both: check `getDisForUnitTestings()`.

Nokel81 replied:
Fair enough. I did this so that we could guarantee (by tests) that things in `common` don't depend on things outside of `common`.

jansav commented on src/common/hotbar-store/__tests__/store.test.ts:
```diff
@@ -0,0 +1,209 @@
+ */
+
+import mockFs from "mock-fs";
+import { AppPaths } from "../../app-paths";
```
Fix imports here

Iku-turso commented on src/common/__tests__/log-search-store.test.ts:
```diff
@@ -0,0 +1,82 @@
+    expect(logSearchStore.occurrences).toEqual([]);
+
+    logSearchStore.onSearch([], "replica");
+    expect(logSearchStore.occurrences).toEqual([]);
```
Multiple expectations per test. Superficial.

Iku-turso commented on src/common/__tests__/log-search-store.test.ts:
```diff
@@ -0,0 +1,82 @@
+/**
```
This file for a technical test could be located right next to implementation. We could have \_\_tests\_\_ for more behavioural stuff. Superficial.

Iku-turso replied:
This stateful store can probably be replaced with a reactive solutions, such as `asyncComputed`. Superficial, for now.

Iku-turso commented on src/common/__tests__/log-search-store.test.ts:
```diff
@@ -0,0 +1,82 @@
+  });
+
+  it("does nothing with empty search query", () => {
+    logSearchStore.onSearch([], "");
```
Calling `onSearch()` feels like stateful solution where there should be reactivity. Superficial.

Iku-turso commented on src/common/__tests__/user-store.test.ts:
```diff
@@ -29,7 +29,7 @@ import { stdout, stderr } from "process";
+import directoryForUserDataInjectable from "../app-paths/directory-for-user-data.injectable";
 import type { ClusterStoreModel } from "../cluster-store/cluster-store";
 import { defaultTheme } from "../vars";
 
```
Notice: criminal global override below.

Iku-turso commented on src/common/catalog-entities/general.ts:
```diff
@@ -19,21 +18,9 @@ export class GeneralEntity extends CatalogEntity<CatalogEntityMetadata, CatalogE
 
-  async onRun() {
+  onRun() {
     navigate(this.spec.path);
```
Notice: explicit side-effect.

Iku-turso commented on src/common/catalog-entities/web-link.ts:
```diff
@@ -25,29 +24,21 @@ export class WebLink extends CatalogEntity<CatalogEntityMetadata, WebLinkStatus,
 
-  async onRun() {
+  onRun() {
     window.open(this.spec.url, "_blank");
```
Notice: explicit side-effect.

Iku-turso commented on src/common/fs/watch-file-path.injectable.ts:
```diff
@@ -2,12 +2,12 @@
-const apiManagerInjectable = getInjectable({
-  instantiate: () => apiManager,
+const watchFilePathInjectable = getInjectable({
+  instantiate: () => watch,
```
This exposes foreign API to our internal modeling. We should wrap this with minimal abstraction. Let's talk about this a bit, otherwise superficial.

Iku-turso commented on src/common/hotbar-store/__tests__/hotbar.test.ts:
```diff
@@ -0,0 +1,207 @@
+      silly: jest.fn(),
+      warn: jest.fn(),
+    };
+    hotbar = new Hotbar(getEmptyHotbar("Default"), { logger });
```
Could be instantiated using di? Superficial.

Iku-turso commented on src/common/hotbar-store/__tests__/hotbar.test.ts:
```diff
@@ -0,0 +1,207 @@
+    hotbar.restackItems(1, 5);
+
+    expect(hotbar.items[5]).toBeTruthy();
+    expect(hotbar.items[5].entity.uid).toEqual("test");
```
Multiple expects. Superficial.

Iku-turso commented on src/common/hotbar-store/__tests__/hotbar.test.ts:
```diff
@@ -0,0 +1,207 @@
+    expect(hotbar.items[5].entity.uid).toEqual("test");
+  });
+
+  it("moves items down", () => {
```
Cryptic test name. Superficial.

Iku-turso commented on src/common/hotbar-store/__tests__/hotbar.test.ts:
```diff
@@ -0,0 +1,207 @@
+/**
```
The naming of tests here and elsewhere would benefit from a naming convention such as gherkin (ie. given, when, then). Let's talk about this elsewhere, otherwise superficial.

Iku-turso commented on src/common/hotbar-store/__tests__/store.test.ts:
```diff
@@ -0,0 +1,209 @@
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+
+import mockFs from "mock-fs";
```
Notice: this beast is very bad, almost always.

Iku-turso commented on src/common/hotbar-store/add-to-active-hotbar.injectable.ts:
```diff
@@ -0,0 +1,27 @@
+  hotbar: IComputedValue<Hotbar>;
+}
+
+function addToActiveHotbar({ hotbar }: Dependencies, entity: CatalogEntity, cellIndex?: number) {
```
Could be lambda. Superficial.

Nokel81 replied:
I prefer the `bind` pattern instead as I find it easier to read.

Iku-turso commented on src/common/hotbar-store/add-to-active-hotbar.injectable.ts:
```diff
@@ -0,0 +1,27 @@
+}
+
+const addToActiveHotbarInjectable = getInjectable({
+  instantiate: (di) => bind(addToActiveHotbar, null, {
```
Here `bind` is an overly exotic and complex pattern. It also limits movement, as it prevents logic in function configuration time. It also hurts aspirations for point-free -programming. Let's not use it.

Iku-turso commented on src/common/hotbar-store/add-to-active-hotbar.injectable.ts:
```diff
@@ -0,0 +1,27 @@
+  hotbar: IComputedValue<Hotbar>;
+}
+
+function addToActiveHotbar({ hotbar }: Dependencies, entity: CatalogEntity, cellIndex?: number) {
```
This should be in new file. Superficial.

Iku-turso commented on src/common/hotbar-store/hotbar.ts:
```diff
@@ -0,0 +1,121 @@
+    this.items[index] = null;
+  });
+
+  restackItems = action((from: number, to: number): void => {
```
Notice: Messy code in many ways. Superficial.

Iku-turso commented on src/common/hotbar-store/hotbar.ts:
```diff
@@ -0,0 +1,121 @@
+  };
+
+  removeItemById = action((uid: string): void => {
+    const index = this.items.findIndex(item => item?.entity.uid === uid);
```
Notice: items should be a `map`.

Iku-turso commented on src/common/hotbar-store/hotbar.ts:
```diff
@@ -0,0 +1,121 @@
+/**
```
Notice: pretty hairy file. Will probably be gone soon, being a store.

Iku-turso commented on src/common/hotbar-store/is-added-to-active-hotbar.injectable.ts:
```diff
@@ -0,0 +1,27 @@
+}
+
+const isItemInActiveHotbarInjectable = getInjectable({
+  instantiate: (di) => bind(isAddedToActiveHotbar, null, {
```
`bind`.

Iku-turso commented on src/common/hotbar-store/remove-from-active-hotbar.injectable.ts:
```diff
@@ -0,0 +1,26 @@
+}
+
+const removeByIdFromActiveHotbarInjectable = getInjectable({
+  instantiate: (di) => bind(removeByIdFromActiveHotbar, null, {
```
`bind`.

Iku-turso commented on src/common/hotbar-store/store.ts:
```diff
@@ -30,7 +31,9 @@ export class HotbarStore extends BaseStore<HotbarStoreModel> {
       },
-      migrations,
+      migrations: ipcMain
+        ? asLegacyGlobalObjectForExtensionApi(hotbarStoreMigrationsInjectable)
```
If this is indeed required, it should be created with other similar globals, instead of inline.

Iku-turso commented on src/common/hotbar-store/store.ts:
```diff
@@ -81,9 +84,9 @@ export class HotbarStore extends BaseStore<HotbarStoreModel> {
       hotbar.items[0] = initialItem;
 
-      this.hotbars = [hotbar];
+      this.hotbars = [new Hotbar(hotbar, { logger })];
```
Notice: responsibility of instantiation is best delegated to DI.

Iku-turso commented on src/common/hotbar-store/store.ts:
```diff
@@ -4,17 +4,18 @@
  */
```
Notice: what a hairy file!

Iku-turso commented on src/common/k8s-api/endpoints/cluster.api.injectable.ts:
```diff
@@ -0,0 +1,14 @@
+import type { ClusterApi } from "./cluster.api";
+
+const clusterApiInjectable = getInjectable({
+  instantiate: (di) => di.inject(apiManagerInjectable).getApi("/apis/cluster.k8s.io/v1alpha1/clusters") as ClusterApi,
```
Hmm. I wonder why the cast? Yet, superficial.

Iku-turso commented on src/common/k8s-api/endpoints/event.api.injectable.ts:
```diff
@@ -0,0 +1,14 @@
+import type { EventApi } from "./event.api";
+
+const eventApiInjectable = getInjectable({
+  instantiate: (di) => di.inject(apiManagerInjectable).getApi("/api/v1/events") as EventApi,
```
Hmm. `getApi` could have a type parameter instead of casting? Superficial.

Iku-turso commented on src/common/k8s-api/endpoints/ingress.api.ts:
```diff
@@ -194,17 +190,14 @@ export class Ingress extends KubeObject {
-let ingressApi: IngressApi;
+export class IngressApi extends KubeApi<Ingress> {
+  constructor(args: SpecificApiOptions<Ingress> = {}) {
+    args.checkPreferredVersion ??= true;
```
Could use default arguments/properties instead of mutation. Superficial.

Iku-turso commented on src/common/k8s-api/endpoints/role.api.ts:
```diff
@@ -4,11 +4,10 @@
 
 export interface Role {
-  rules: {
+  rules?: {
```
Hmm. In general, perhaps we tolerate nullable types a bit too much in the codebase, as they contribute to complexity quite a bit. Just a thought.

Nokel81 replied:
I agree, however we must respect the fact that a lot of these are because they are nullable from kube itself.

Iku-turso commented on src/common/k8s-api/api-manager.injectable.ts:
```diff
@@ -0,0 +1,256 @@
+/**
```
This whole file could be made redundant by making the injectable API's also declare `setup()`. This would also comply to OCP nicely. Superficial, but very important.

Nokel81 replied:
I can do that in a follow up PR

Iku-turso commented on src/common/logger/create-prefixed-logger.injectable.ts:
```diff
@@ -0,0 +1,32 @@
+
+const createPrefixedLoggerInjectable = getInjectable({
+  instantiate: () => bind(createPrefixedLogger, null, {
+    // TODO make injectable
```
Also: tag with `causesSideEffects: true`. Superficial.

Nokel81 replied:
How does it cause side effects?

Nokel81 replied:
Or more specifically, what does that flag do/mean?

Iku-turso commented on src/common/migrations/cluster-store/migrations.injectable.ts:
```diff
@@ -0,0 +1,26 @@
+// Cluster store migrations
+
+import { joinMigrations } from "../helpers";
+import snap from "./snap";
```
Could have better name for readability. Superficial.

Iku-turso commented on src/common/utils/tar.ts:
```diff
@@ -47,6 +47,8 @@ export function readFileFromTar<R = Buffer>({ tarPath, filePath, parseJson }: Re
 export async function listTarEntries(filePath: string): Promise<string[]> {
   const entries: string[] = [];
 
+  // The typings of this is wrong, we do need to await
```
Weird. Superficial.

Iku-turso commented on src/common/utils/unique-id.injectable.ts:
```diff
@@ -0,0 +1,13 @@
+import { getInjectable, lifecycleEnum } from "@ogre-tools/injectable";
+import { uniqueId } from "lodash";
+
+const uniqueIdInjectable = getInjectable({
```
This should be tagged with `causesSideEffects: true`. Superficial, yet very notable.

Iku-turso commented on src/extensions/lens-renderer-extension.ts:
```diff
@@ -53,31 +60,35 @@ export class LensRendererExtension extends LensExtension {
    */
   addCatalogCategoryFilter(fn: CategoryFilter): Disposer {
-    const dispose = catalogCategoryRegistry.addCatalogCategoryFilter(fn);
+    const dispose = once(() => this.catalogCategoryFilters.remove(fn));
```
Redundant `once`. Superficial.

Nokel81 replied:
Not really, I think it helps prevent bugs from the extensions side.

Iku-turso commented on src/main/catalog/extension-sourced-entities.injectable.ts:
```diff
@@ -0,0 +1,38 @@
+
+function getExtensionSourcesComputed({ extensions }: Dependencies): IComputedValue<CatalogEntity[]> {
+  return computed(() => (
+    Array.from(
```
Use functional composition or collection chaining to normalize nested code, and make it left-to-right/top-to-bottom -readable.

Iku-turso commented on src/main/protocol-handler/router.injectable.ts:
```diff
@@ -0,0 +1,22 @@
+    new LensProtocolRouterMain({
+      extensionLoader: di.inject(extensionLoaderInjectable),
+      extensionsStore: di.inject(extensionsStoreInjectable),
+      ensureMainWindow: di.inject(ensureMainWindowInjectable),
```
LensProtocolRouterMain does not need to know about this, it only needs to wait for it to be "ensured" before instantiation? Superficial, interesting.

Iku-turso commented on src/main/index.ts:
```diff
@@ -25,49 +22,51 @@ import type { LensExtensionId } from "../extensions/lens-extension";
-di.runSetups().then(() => {
-  injectSystemCAs();
+async function main() {
+  console.log("1");
```
Forgotten debug-logging.

Iku-turso commented on src/main/index.ts:
```diff
@@ -260,7 +250,7 @@ di.runSetups().then(() => {
       kubeConfigSyncManager.startSync();
 
-      startUpdateChecking();
+      di.inject(startUpdateCheckingInjectable)();
```
Hmm... This could be a setuppable? Superficial, important.

Iku-turso commented on src/renderer/catalog/on-run.injectable.ts:
```diff
@@ -0,0 +1,13 @@
+import catalogEntityRegistryInjectable from "./entity-registry.injectable";
+
+const onRunInjectable = getInjectable({
+  instantiate: (di) => di.inject(catalogEntityRegistryInjectable).onRun,
```
Hmm. I'm thinking if roughly this kind of delegation could be used elsewhere as well to cut wings from more `reactions` and `autoruns`. Mental note.

Iku-turso commented on src/renderer/catalog/render-context-menu-item.injectable.tsx:
```diff
@@ -0,0 +1,65 @@
+export type RenderEntityContextMenuItem = (display: "title" | "icon") => (menuItem: CatalogEntityContextMenu, index: number) => React.ReactNode;
+
+const renderEntityContextMenuItemInjectable = getInjectable({
+  instantiate: (di) => bind(registerEntityContextMenuItem, null, {
```
`bind`

Iku-turso commented on src/renderer/components/+catalog/catalog-menu.tsx:
```diff
@@ -42,39 +40,51 @@ function Item(props: TreeItemProps) {
+  </StylesProvider>
+));
+
+export const CatalogMenu = withInjectables<Dependencies, CatalogMenuProps>(NonInjectedCatalogMenu, {
```
No dependencies?

Nokel81 replied:
Ah, oops. Past in the one dependency as a prop by accident.

Nokel81 committed:
Complete a majority of the rest of DI

- All KubeApi and KubeObjectStore's

- All of Catalog

- ApiManager

- KubeObjectDetailRegistry

- All BaseStore's and their migrations

- LensProxy and Router

- WindowManager

- App Menu

- Tray Icon

- ThemeStore

Signed-off-by: Sebastian Malton <sebastian@malton.name>
